src/iov.c

```diff
@@ -63,24 +63,39 @@ uint64_t ofi_copy_iov_buf(const struct iovec *iov, size_t iov_count, uint64_t io
 void ofi_consume_iov_desc(struct iovec *iov, void **desc,
 			  size_t *iov_count, size_t to_consume)
 {
-	size_t i;
+	struct iovec *cur_iov;
+	void **cur_desc;
+
+	assert(to_consume && *iov_count);
+	assert(to_consume <= ofi_total_iov_len(iov, *iov_count));
+
+	if (*iov_count == 1 && iov[0].iov_len > to_consume) {
+		iov[0].iov_len -= to_consume;
+		return;
+	} else if (*iov_count == 1) {
+		*iov_count = 0;
+		return;
+	}
+
+	cur_iov = iov;
+	cur_desc = desc;
+	while (*iov_count && to_consume >= cur_iov->iov_len) {
+		to_consume -= cur_iov->iov_len;
+		cur_iov++;
+		cur_desc++;
+		*iov_count -= 1;
+	}
+	if (*iov_count == 0)
+		return;
 
-	if (*iov_count == 1)
-		goto out;
+	cur_iov->iov_len -= to_consume;
+	cur_iov->iov_base = (uint8_t *)iov[0].iov_base + to_consume;
```

Comment:
FWIW, this code is updating the iov that will become iov[0].  It's just being done prior to the memmove below.  If this code were moved to after the if statement below, cur_iov could be replaced with iov[0], which is what the original code had.  The iov_count == 1 check above could then use a goto to fall to it, also similar to the original code.  The missing piece from the original code was setting iov_count = 0 when fully consuming, which is now handled above.